### PR TITLE
Fix #94: Write command line tool to validate SARIF files

### LINKED ISSUE #94: Write command line tool to validate SARIF files against schema

### PR DESCRIPTION
**Pull request by lgolding**
@michaelcfanning

### CODE REVIEW

File: src/Tools/SarifValidator/JSONError.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿namespace Microsoft.CodeAnalysis.Sarif.SarifValidator
```
Comment (lgolding) [inline src/Tools/SarifValidator/JSONError.cs]: This file copied from the json-validator repo.

File: src/Tools/SarifValidator/ConsoleUtil.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Globalization;
+
+namespace Microsoft.CodeAnalysis.Sarif.SarifValidator
+{
+    internal static class ConsoleUtil
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/ConsoleUtil.cs]: Why do people use 'Util'? is there a problem with adding three more letters to this term so that it comprises a complete word? :)

File: src/Tools/SarifValidator/Validator.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            string instanceText = File.ReadAllText(instanceFilePath);
+            string schemaText = File.ReadAllText(schemaFilePath);
+
+            return Validate(instanceText, schemaText);
+        }
+
+        private static IEnumerable<JSONError> Validate(string instanceText, string schemaText)
+        {
+            List<JSONError> errors = new List<JSONError>();
+
```
Comment (lgolding) [inline src/Tools/SarifValidator/Validator.cs]: This logic modeled on the code in the json-validator repo.

File: src/Tools/SarifValidator/JSONError.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    public enum JSONErrorLocation
+    {
+        InstanceDocument,
+        Schema
+    }
+
+    public class JSONError
+    {
+        public JSONErrorLocation Location { get; set; }
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/JSONError.cs]: This casing is wrong, s/be JsonErrorLocation.

File: src/Tools/SarifValidator/packages.config [hunk 5]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="CommandLineParser" version="2.0.275-beta" targetFramework="net46" />
+  <package id="Microsoft.CodeAnalysis.SarifValidationLibraries" version="1.0.0" targetFramework="net46" />
```
Comment (lgolding) [inline src/Tools/SarifValidator/packages.config]: This is the package I made yesterday.

File: src/Tools/SarifValidator/Options.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+using CommandLine;
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/Options.cs]: Where are the MIT headers?

File: src/Tools/SarifValidator/ConsoleUtil.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        internal static void WriteError(string message, params object[] args)
+        {
+            WriteColor(ConsoleColor.Red, message, args);
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/ConsoleUtil.cs]: in my experience colorized console output is not a great idea. Will your green results show up well against a black background? What about people (like me) who uses several different color schemes, for readability and to also keep track of different cmd environments (my CBT shell looks different than my VS env, for example). Also, someone with red/green color blindness will not benefit from this choice. Just my 2 cents on the topic.

File: src/Tools/SarifValidator/Program.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            Version version = typeof(Program).Assembly.GetName().Version;
+
+            Console.WriteLine(string.Format(CultureInfo.CurrentCulture, Resources.Banner, version));
+            Console.WriteLine(Resources.Copyright);
+            Console.WriteLine();
+        }
+
+        private static void ShowOptions(Options options)
+        {
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/Program.cs]: command line library has a built in options reporting feature, yes? do we need to reauthor one?

File: src/Tools/SarifValidator/ConsoleUtil.cs (hunk 2 above)
Comment (lgolding) [inline src/Tools/SarifValidator/ConsoleUtil.cs]: Fixed. ("ConsoleUtilities", plural)

File: src/Tools/SarifValidator/JSONError.cs (hunk 4 above)
Comment (lgolding) [inline src/Tools/SarifValidator/JSONError.cs]: Fixed. Note that the libraries have several types that are named "JSONXxx"; can't change that.

File: src/Tools/SarifValidator/SarifValidator.csproj [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  <PropertyGroup>
+    <SignAssembly>true</SignAssembly>
+  </PropertyGroup>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., build.props))\build.props" />
+  <PropertyGroup>
+    <!-- Override the TargetFrameworkVersion specified in build.props. This project must target
+         .NET 4.6 because the assemblies from the SarifValidationLibraries NuGet package,
+         upon which this project depends, target that version. -->
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <AssemblyOriginatorKeyFile>..\..\..\GeneratedKey.snk</AssemblyOriginatorKeyFile>
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/SarifValidator.csproj]: Why are we signing this tool? Will we ship it in the nuget package at some point?

File: src/Tools/SarifValidator/Options.cs (hunk 6 above)
Comment (lgolding) [inline src/Tools/SarifValidator/Options.cs]: Fixed.

File: src/Tools/SarifValidator/Validator.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            AddValidationErrors(tree.ValidationIssues, errors);
+
+            return errors;
+        }
+
+        private static void AddSyntaxErrors(JSONDocument document, JSONErrorLocation location, List<JSONError> errors)
+        {
+            foreach (Tuple<JSONParseItem, JSONParseError> parserError in document.GetContainedParseErrors())
+            {
+                errors.Add(new JSONError
```
Comment (michaelcfanning) [inline src/Tools/SarifValidator/Validator.cs]: I don't understand why you didn't author your output as SARIF, in the ultimate dogfooding exercise. :)

File: src/Tools/SarifValidator/SarifValidator.csproj (hunk 9 above)
Comment (lgolding) [inline src/Tools/SarifValidator/SarifValidator.csproj]: Didn't see any harm in it.

File: src/Tools/SarifValidator/Program.cs (hunk 8 above)
Comment (lgolding) [inline src/Tools/SarifValidator/Program.cs]: This isn't a "help" feature, it reports the values of the options you selected. But I can remove it if you don't think it's valuable.

File: src/Tools/SarifValidator/ConsoleUtil.cs (hunk 7 above)
Comment (lgolding) [inline src/Tools/SarifValidator/ConsoleUtil.cs]: Ok, I'll remove.

File: src/Tools/SarifValidator/ConsoleUtil.cs (hunk 2 above)
Comment (lgolding) [inline src/Tools/SarifValidator/ConsoleUtil.cs]: ... but then removed per comment below.

File: src/Tools/SarifValidator/Program.cs (hunk 8 above)
Comment (lgolding) [inline src/Tools/SarifValidator/Program.cs]: Removed.

File: src/Tools/SarifValidator/Validator.cs (hunk 10 above)
Comment (lgolding) [inline src/Tools/SarifValidator/Validator.cs]: OK, will do in a separate PR.

Comment (michaelcfanning) [inline src/Tools/SarifValidator/Validator.cs]: Terrific! Then this tool's output can also be verified by the tool and now my brain hurts.

**Comment by lgolding** [comment]
Verbal +1 from @michaelcfanning
